src/main/java/io/github/ossgang/commons/observable/DispatchingObservable.java

```diff
@@ -43,13 +45,31 @@ protected DispatchingObservable() {
     public Subscription subscribe(Consumer<? super T> listener, SubscriptionOption... options) {
         Set<SubscriptionOption> optionSet = new HashSet<>(Arrays.asList(options));
         if (optionSet.contains(WEAK)) {
-            listeners.put(new WeakConsumer<>(listener, listeners::remove), optionSet);
+            addListener(new WeakConsumer<>(listener, this::removeListener), optionSet);
         } else {
-            listeners.put(listener, optionSet);
+            addListener(listener, optionSet);
         }
         return new ObservableSubscription(listener);
     }
 
+    private void addListener(Consumer<? super T> listener, Set<SubscriptionOption> subscriptionOptions) {
+        listeners.put(listener, subscriptionOptions);
+        if (listenerCount.getAndIncrement() == 0) {
+            firstListenerAdded();
+        }
+    }
+
+    protected void firstListenerAdded() { /* no op */ }
+
+    private void removeListener(Consumer<? super T> listener) {
+        listeners.remove(listener);
```

Comment:
Should we check if the listener was effectively removed before counting down?

Reply:
Well, I would say this is an invalid situation - so I'd throw IllegalStateException if the listener was not actually removed.

Reply:
Is ok ... not so sure though ... I think one also could ignore it...(and just not decrement).... (There could be situations where one just would like to make sure that the listeners is not there and one would 'proactively' remove them .... as one has no way to check if its there ...

Reply:
Ok, that's a valid point - someone could keep a Subscription object - if we throw the user must make sure to only call unsubscribe once (which can be a bit tedious in a multithreaded environment), if we are lenient one can just try to call unsubscribe ...